### PR TITLE
AutoLoader for - 2701 speed bumps

### PR DESCRIPTION
**Pull request by anthony2727**
Hi Guys, 

Just making a contribution to this project. Please, read comments. 

I'm implementing an autoloading approach so It isn't necessary to include or require any php file manually. This solution is also compatible with the standard specified by dependency managers like Composer. 

I really appreciate any feedback and I hope you find it useful. 

Thanks,

### CODE REVIEW

File: inc/SpeedBumps.php [hunk 1]
```diff
@@ -1,31 +1,53 @@
 <?php
+namespace SpeedBumps\Inc;
 
-class Speed_Bumps {
+class SpeedBumps {
```
Comment (noppanit) [inline inc/SpeedBumps.php]: @anthony2727 just out of curiosity. Why did you change the filename?

**Comment by noppanit** [comment]
@goldenapples what do you think of this approach?

**Comment by anthony2727** [comment]
Hi @noppanit, 

Thank you for asking. 

The "WHY" is commented in the loader.php. 
I commented: 
"Notes:  
The fileName and class should have the same name. That's why for example, I changed the class-speed-bumps.php name to SpeedBumps so when I create a new instance $speedBumps = new SpeedBumps, the $file variable from line 27 will construct itself taking the name of the class which matches with the name of the file."

Any further question, just let me know.

File: inc/SpeedBumps.php [hunk 2]
```diff
@@ -1,31 +1,53 @@
 <?php
+namespace SpeedBumps\Inc;
```
Comment (goldenapples) [inline inc/SpeedBumps.php]: I'm not a big fan of "Inc" being part of the name space path. It doesn't add any information. All of the PHP code included in this plugin will live in `/inc/`. Can that path part be added in by the autoloader so we don't have to put everything in an `\Inc` namespace?

**Comment by goldenapples** [comment]
> @goldenapples what do you think of this approach? 

I like it, and it makes plugin structure much simpler. I would just like to make the namespace organizational structure a little more semantic as you've done in #1. (This will also mean organizing the class files into a directory structure that matches the namespace hierarchy.)

So just to be clear, if a theme wants to define an additional class in one of these namespaces, they just need to define it before using it in registering a speed bump, and their version of the class will be used, and the plugin won't try to autoload its own.

File: inc/SpeedBumps.php (hunk 2 above)
Comment (anthony2727) [inline inc/SpeedBumps.php]: Absolutely.

File: inc/SpeedBumps.php (hunk 1 above)
Comment (danielbachhuber) [inline inc/SpeedBumps.php]: We should keep the name as `Speed_Bumps`, per WP coding standards.

File: inc/SpeedBumps.php [hunk 3]
```diff
@@ -1,31 +1,53 @@
...
+class SpeedBumps {
 	private static $instance;
 	private static $_speed_bumps_args = array();
 
 	public static function get_instance() {
 		if ( ! isset( self::$instance ) ) {
 
-			self::require_files();
-			self::$instance = new Speed_Bumps;
+			//self::require_files();
```
Comment (danielbachhuber) [inline inc/SpeedBumps.php]: You can remove this comment.

File: inc/SpeedBumpsElementConstraint.php [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<?php
```
Comment (danielbachhuber) [inline inc/SpeedBumpsElementConstraint.php]: This filename should be `class-element-constraint.php`

Comment (anthony2727) [inline inc/SpeedBumpsElementConstraint.php]: So, classname should be Name_Of_The_Class and the filename: name-of-the-file.php? @danielbachhuber

Comment (danielbachhuber) [inline inc/SpeedBumpsElementConstraint.php]: https://make.wordpress.org/core/handbook/coding-standards/php/#naming-conventions

Comment (anthony2727) [inline inc/SpeedBumpsElementConstraint.php]: Cool, I can make some changes to adapt the autoloader.

**Comment by anthony2727** [comment]
@goldenapples I didn't get your point with: "So just to be clear, if a theme wants to define an additional class in one of these namespaces, they just need to define it before using it in registering a speed bump, and their version of the class will be used, and the plugin won't try to autoload its own."

**Comment by noppanit** [comment]
Closing in favour of https://github.com/fusioneng/speed-bumps/pull/12
